homeassistant/components/geonetnz_quakes/__init__.py

```diff
@@ -96,6 +125,117 @@
     manager = hass.data[DOMAIN][FEED].pop(config_entry.entry_id)
     await manager.async_stop()
 
-    await hass.config_entries.async_forward_entry_unload(config_entry, "geo_location")
+    for domain in COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_unload(config_entry, domain)
+        )
 
     return True
+
+
+class GeonetnzQuakesFeedEntityManager:
+    """Feed Entity Manager for GeoNet NZ Quakes feed."""
+
+    def __init__(self, hass, config_entry, radius_in_km, unit_system):
+        """Initialize the Feed Entity Manager."""
+        self._hass = hass
+        self._config_entry = config_entry
+        coordinates = (
+            config_entry.data[CONF_LATITUDE],
+            config_entry.data[CONF_LONGITUDE],
+        )
+        websession = aiohttp_client.async_get_clientsession(hass)
+        self._feed_manager = GeonetnzQuakesFeedManager(
+            websession,
+            self._generate_entity,
+            self._update_entity,
+            self._remove_entity,
+            coordinates,
+            mmi=config_entry.data[CONF_MMI],
+            filter_radius=radius_in_km,
+            filter_minimum_magnitude=config_entry.data[CONF_MINIMUM_MAGNITUDE],
+            filter_time=DEFAULT_FILTER_TIME_INTERVAL,
+            status_callback=self._status_update,
+        )
+        self._config_entry_id = config_entry.entry_id
+        self._scan_interval = timedelta(seconds=config_entry.data[CONF_SCAN_INTERVAL])
+        self._unit_system = unit_system
+        self._track_time_remove_callback = None
+        self._status_info = None
+        self.listeners = []
+
+    async def async_init(self):
+        """Schedule initial and regular updates based on configured time interval."""
+
+        for domain in COMPONENTS:
+            self._hass.async_create_task(
+                self._hass.config_entries.async_forward_entry_setup(
+                    self._config_entry, domain
+                )
+            )
+
+        async def update(event_time):
+            """Update."""
+            await self.async_update()
+
+        self._track_time_remove_callback = async_track_time_interval(
+            self._hass, update, self._scan_interval
+        )
+
+        async def component_loaded(event):
+            """Handle a new component loaded."""
+            component = event.data.get(ATTR_COMPONENT)
+            if component == DOMAIN:
+                await self.async_update()
+
+        self._hass.bus.async_listen(EVENT_COMPONENT_LOADED, component_loaded)
```

Comment:
Why do we need this?

Reply:
We have `setup.async_when_setup`. But please explain the problem first.

Reply:
I added this as a trigger for the first update of the feed entity manager. Without this trigger the first update would only happen after the first update interval. And, because manager and platforms communicate via signals, I wanted to make sure that all platforms in this component have been setup before the first update is triggered.

Reply:
But the component of this domain will already be setup at this point. We want to check that the platformd are ready. That's not the same.

Reply:
I see. I thought that `EVENT_COMPONENT_LOADED` is only triggered after the component and all platforms have been setup.

Is there a good way to wait for all platforms to have completed their `async_setup_entry` method? In `GeonetnzQuakesFeedEntityManager#async_init` I loop through the platforms and create an async task to set up each one, but unfortunately this approach gives away control over when these tasks are completed.

Reply:
I just saw that I could hold on to each task and add a listener via `add_done_callback`. But I could find a single example, so not sure if this is good practice.

Reply:
I like that. Where is the existing example?

Reply:
I found it here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/homematicip_cloud/hap.py#L131

It's not exactly the same as my challenge. I would basically add a callback to each job. In that callback I would have a checklist that for each platform ticks off if it's setup, and after all have been setup call the manager's update method.

Reply:
Would it be enough to call the update after each platform setup in `async_setup_entry` for the platform. Does it matter if the update method gets called multiple times?

Reply:
Yes, that might actually be a good compromise. It would only happen once at startup. I'll give it a try.

Reply:
Just realised: If the order coincidentally turns out to be as follows:

1. sensor platform setup
2. update manager
3. geolocation platform setup
4. update manager

Then the first update of the manager will try to signal creating new geolocation entities, and in the second update assume that they exist and signal an update. But because they haven't been set up in the first place, they will never be created.

Also, my unit tests become unpredictable, because I am setting up a certain configuration, and need to assert that the setup was successful. But because there is no single event that tells the test that the setup was successful, the asserts sometimes work and sometimes don't.

I may need to go back and control the order of the platform setup to make this all work.

Reply:
Better idea: The geolocation setup triggers an update of the manager. But the sensor won't update the manager, but instead just updates itself, i.e. retrieves the current status from the manager, or subscribe to status changes and will be updated as soon as the geolocation platform has triggered the update.